autograd/numpy/numpy_vjps.py

```diff
@@ -535,17 +535,53 @@ def match_complex(target, x):
     else:
         return x
 
+@notrace_primitive
+def _needs_broadcast(x, target_meta):
+    target_shape, _, _, target_iscomplex = target_meta
+    return (onp.shape(x) != target_shape
+            or (target_iscomplex != onp.iscomplexobj(x)))
+
+def broadcast(x, target_meta, broadcast_idx=0):
+    if _needs_broadcast(x, target_meta):
+        return _broadcast(x, target_meta, broadcast_idx)
+    return x
+
+@primitive
+def _broadcast(x, target_meta, broadcast_idx=0):
+    target_shape, _, _, target_iscomplex = target_meta
+    x = onp.broadcast_to(x, target_shape)
+    if target_iscomplex and not onp.iscomplexobj(x):
+        x = x + 0j  # TODO(mattjj): this might promote the dtype
+    return x
+
+def grad_broadcast(ans, x, target_meta, broadcast_idx=0):
+    meta = anp.metadata(x)
+    return lambda g: _unbroadcast(g, meta, broadcast_idx)
+defvjp(_broadcast, grad_broadcast)
+
 def unbroadcast(x, target_meta, broadcast_idx=0):
-    target_shape, target_ndim, dtype, target_iscomplex = target_meta
-    while anp.ndim(x) > target_ndim:
-        x = anp.sum(x, axis=broadcast_idx)
+    if _needs_broadcast(x, target_meta):
+        return _unbroadcast(x, target_meta, broadcast_idx)
+    return x
+
+@primitive
+def _unbroadcast(x, target_meta, broadcast_idx=0):
+    target_shape, target_ndim, _, target_iscomplex = target_meta
+    x_shape = onp.shape(x)
+    while onp.ndim(x) > target_ndim:
+        x = onp.sum(x, axis=broadcast_idx)
```

Comment:
I was wondering if we should replace the above two lines with:
```python
x = onp.sum(x, axis=range(broadcast_idx, broadcast_idx + onp.ndim(x) - target_ndim))
```
or similar. Am I right that only calling `sum` once might lead to better performance?

Reply:
I briefly tried something like that (though if `broadcast_idx` is -1, which is the only nonzero use case I noticed in the code, then I think we want something different) and it didn't seem to make a speed difference, so I dropped it. Now is a good time to make sure it's performant, though!

Reply:
Doing a few timings it looks like there is a benefit for small arrays but it's not massive:
```python
In [15]: a = np.ones((5, 5, 5))

In [16]: %timeit np.sum(a, axis=(0, 1))
5.38 µs ± 112 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [17]: %timeit x = np.sum(a, axis=0); x = np.sum(x, axis=0)
8.62 µs ± 124 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
and for slightly bigger arrays it's the other way round (maybe I've made some mistake?):
```python
In [18]: a = np.ones((50, 50, 50))

In [19]: %timeit np.sum(a, axis=(0, 1))
118 µs ± 930 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [20]: %timeit x = np.sum(a, axis=0); x = np.sum(x, axis=0)
81.6 µs ± 1.54 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Reply:
Wow, I got similar timings. That seems weird for the bigger arrays...